### ISSUE TITLE
Multi-company invoices and revenues create errors

### ISSUE THREAD

**Issue by iahmadaj** [akaunting/akaunting#804]
Please, submit only real issues. Use the forum for support, feature requests, proposals, new versions, help etc. https://akaunting.com/forum

### Steps to reproduce the issue
1- Create a new company and switch to that company
2- Go to incomes->invoices or incomes->revenues
3- Click on add new and it will show "Whoops, looks like something went wrong." 
4- Also an error when you try to go to general settings for this 2nd company

### Expected result
1- It should show a form to add invoices or Revenues
2- For settings it should show general settings form

### Actual result
Displays an error page with "Whoops, looks like something went wrong." error

### System information (Akaunting, PHP versions)
Akaunting_1.3.14-Stable version, and 

PHP Version 7.1.27

### Additional comments
After debugging I noticed that for 2nd company the settings table is not being populated with some required variable for invoices and revenues. After I changed the code and added those settings variables it started working as expected. 

I just want to know if I am doing anything wrong while adding new company or during company management do I need to do anything extra which I am missing.

**Comment by vplaya** [comment]
This is a serious problem with me too and it will be appreciated if we can be assisted here.

**Comment by cuneytsenturk** [comment]
thank for reporting @iahmadaj

**Comment by mgschwan** [comment]
I found a partial solution for this problem.

This only happens if there are no offline payment methods configured, which is the case when a new company is created.

However, creating a new offline payment method via Settings -> Offline Payments does not work either if the list of offline payments is empty.

A workaround is to add the offline payments directly in the database via:

insert into hdu_settings ( `company_id`, `key`, `value` ) VALUES ( 2 , 'offlinepayment.methods', '[{"code":"offlinepayment.cash.1","name":"Cash","order":"1","description":null},{"code":"offlinepayment.bank_transfer.2","name":"Bank Transfer","order":"2","description":null}]' );

Where hdu_settings is the name of the settings table in your database (the first three letters seem to be randomly generated at the time of the installation)

This has to be done for every company, and the company_id (in this example the id is 2 ) has to be changed accordingly

**Comment by vplaya** [comment]
Editing the database to solves this seem to be challenging for some of us, is there anyone who could assist in a layman terms please. I access my data base via phpmyadmin. thanks in advance.

**Comment by mgschwan** [comment]
This should not be the official "solution" but for those that are using phpmyadmin go to the settings table of your akaunting database. It's called xxx_settings where xxx is a random 3 letter prefix string that was set at the time of installation.

Click on Insert.

In the company_id field enter the id of the company you are trying to edit. If it is the second company you created then the company id should be 2.

In the key field enter `offlinepayment.methods`

In the value field paste the following json string
`[{"code":"offlinepayment.cash.1","name":"Cash","order":"1","description":null},{"code":"offlinepayment.bank_transfer.2","name":"Bank Transfer","order":"2","description":null}]`

Click Go and it should be done.

**Comment by vplaya** [comment]
Thanks Michael,

Your assistance is highly appreciated.

Best Regards,
Thebe

**Comment by vplaya** [comment]
Hi Michael,

The error seems to be still popping up when I click on setting-general. Any ideas?

Are the function fields have to be left blank? please see the attachment
![snapshot](https://user-images.githubusercontent.com/42912594/55277397-50aefd80-5308-11e9-8bcb-ac506968fafb.png)

**Comment by mgschwan** [comment]
Yes that's fine the id is automatically filled and the rest is ok.
But your value seems to be missing the bracket ]  at the end.

The string should end with }] but in your case it's only }

**Comment by vplaya** [comment]
Thank you. Let me try again.

**Comment by vplaya** [comment]
I tried carefully with no luck. I can create a new invoice though but "setting-general and Income-revenue-add new" still throws the "oops"error.
![snapshot](https://user-images.githubusercontent.com/42912594/55277871-75f23a80-530d-11e9-93f8-29120473308a.png)

**Comment by mgschwan** [comment]
Can you export the settings table as a textfile and send it to me. 

BR
Michael 

Am 30.03.2019 16:12, schrieb vplaya:

> I tried carefully with no luck. I can create a new invoice though but "setting-general and Income-revenue-add new" still throws the "oops"error.
> [1] 
> 
> --
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub [2], or mute the thread [3].

 

Links:
------
[1]
https://user-images.githubusercontent.com/42912594/55277871-75f23a80-530d-11e9-93f8-29120473308a.png
[2]
https://github.com/akaunting/akaunting/issues/804#issuecomment-478254787
[3]

**Comment by vplaya** [comment]
Sure Ill send it to your email shortly.

**Comment by iahmadaj** [comment]
Hi vplaya,

I have found out a solution without going to phpmyadmin and updating the database manually. 

Following are the steps we must follow to use Akaunting Software effectively with multi-company feature 
 
1- Install the software on your server and create initial company and admin account. For single company this would be enough to use the software you can also create more users as per their roles e.g managers, customers etc. 

2- If you want to use multi-company feature of the Akaunting software you must follow following steps to avoid errors 

- On top left corner of the admin panel under Akaunting logo you will see name of your company and under that there is a drop down titled switched. Click that drop down and go to “Manage Companies” 
 
- While at Companies management page you will see “Add New” button use that to add a new company. Clicking on Add New button will open a simple new company form, filling and submitting that form will create new company. 
 
- After creating new company, you have to keep in mind that you still can’t use it until necessary data is not added which can be added after switching to this new company and by using settings menu. 
 
- Once you are switched to new company click on settings menu from settings menu go to Offline Payments menu and it will open a form add this data because it is required to manage accounts of this new company you can copy this data from Initial company which was setup at the time of installation of the software.  
 
- After adding Offline Payments data still, you cannot use this new company smoothly. Because still there is some required info to manage account for this company. That required information can only be added by the admin user of this new company so. After creating a new company, you must create an admin user for this company and new admin user can be added using users management.  

- Once a user is created for the company now that user needs to log in using provided credentials and complete the company setup. Once this new user will login a company setup wizard will show up that will be like the one showed up while installing the software first time. This user must finalize the setup for this new company by saving first step and next steps can be skipped. After finalizing the setup now company is ready to be used and managed.  

These above steps needs to be followed to manage multi-company accounts using Akaunting software.

**Comment by srmrox** [comment]
Looks like steps that can be automated (with user input required for necessary fields).

I'll see if I can dig into the code and refactor something; I'm new to the software, though - might take some time to get a hang of things first.

**Comment by vplaya** [comment]
Hi iahmadaj, 

You might have overlooked Michael's comment above. Adding offline info directly from the settings menu doesn't work, hence he suggested the database route which it is the only way currently.

The minute you hit save, whatever data you added doesn't get saved.

Though the database route work for adding the offline payment data, the oops error still shows up when you try to save general settings.

At this stage one can just hope that the Akaunting developers have noticed this and that it will be sorted from the next release.

**Comment by iahmadaj** [comment]
Hi vplaya,

I actually have tested the steps. Well directly general settings are not available but Settings->offlinepayments is available so you can go to that page and add offline payments

After even adding offline payments from there it still doesn't work because still some more settings needed which can be added by admin user for 2nd company and that only can happen by creating an admin user and logging in through that admin user and completing the company setup wizard this way it has worked for me and I have tested in database it adds all necessary settings. Hope it will help non technical users.

Note in offline payments you need to add two entries same like those in base company.

Regards
Irfan Ahmad

**Comment by vplaya** [comment]
Hi  iahmadaj,

After you add the first cash offline data and hit save, does it get saved
before you add the second offline data?

Because from my side as soon as you try to save the first offline info it
stays blank so I never tried to add the second required info.

**Comment by iahmadaj** [comment]
Hi vplaya,

Yes it gets saved for me and I have tried the same two three times and it seems to work fine on saving data for offline payments. I am using Akaunting_1.3.14-Stable version.

Regards
Irfan

**Comment by vplaya** [comment]
Oh I see, I'm on 1.3.15.

**Comment by iahmadaj** [comment]
I can see Mr. Micheal has created a ticket for the issue so we can hope it will be resolved soon. Happy opensourcing :)

**Comment by vplaya** [comment]
Yeah sure:)
